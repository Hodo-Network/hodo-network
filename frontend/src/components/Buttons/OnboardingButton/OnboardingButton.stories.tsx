import { Story, Meta } from "@storybook/react";

import {
  PureOnboardingButton,
  PureOnboardingButtonProps,
} from "./OnboardingButton";

export default {
  title: "Components/Buttons/OnboardingButton",
  component: PureOnboardingButton,
} as Meta;

const Template: Story<PureOnboardingButtonProps> = (args) => (
  <PureOnboardingButton {...args} />
);

export const Default = Template.bind({});
Default.args = {};

export const Disabled = Template.bind({});
Disabled.args = {
  disabled: true,
};
