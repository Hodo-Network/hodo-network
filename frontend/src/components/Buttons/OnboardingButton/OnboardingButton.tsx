import Button from "../../../base/Button";

export interface PureOnboardingButtonProps {
  label?: string;
  disabled?: boolean;
  onConnect?: () => void;
}

export const PureOnboardingButton = ({
  label = "Connect",
  disabled = false,
  onConnect,
}: PureOnboardingButtonProps) => {
  return (
    <Button color="primary" onClick={onConnect} disabled={disabled}>
      {label}
    </Button>
  );
};
