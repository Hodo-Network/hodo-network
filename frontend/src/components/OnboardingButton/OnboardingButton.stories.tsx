import { Story, Meta } from "@storybook/react";

import {
  PureOnboardingButton,
  PureOnboardingButtonProps,
} from "./OnboardingButton";

export default {
  title: "Components/OnboardingButton",
  component: PureOnboardingButton,
} as Meta;

const Template: Story<PureOnboardingButtonProps> = (args) => (
  <PureOnboardingButton {...args} />
);

export const Default = Template.bind({});
Default.args = {};
