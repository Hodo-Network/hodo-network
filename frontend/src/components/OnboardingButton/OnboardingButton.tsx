import Button from "../Button";

export interface PureOnboardingButtonProps {
  label?: string;
  disabled?: boolean;
  onConnect?: () => void;
}

export const PureOnboardingButton = ({
  label = "Connect",
  disabled = false,
  onConnect,
}: PureOnboardingButtonProps) => {
  return (
    <Button onClick={onConnect} disabled={disabled}>
      {label}
    </Button>
  );
};
