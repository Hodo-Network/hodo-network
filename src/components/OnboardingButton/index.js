import React, { useEffect, useState, useRef } from "react";
import MetaMaskOnboarding from "@metamask/onboarding";
import { TEXT_INSTALL_METAMASK, TEXT_CONNECT } from "../../constants/text";

export default function OnboardingButton() {
  const [buttonText, setButtonText] = useState(TEXT_INSTALL_METAMASK);
  const [isDisabled, setDisabled] = useState(false);
  const [accounts, setAccounts] = useState([]);
  const onboarding = useRef();

  useEffect(() => {
    if (!onboarding.current) {
      onboarding.current = new MetaMaskOnboarding();
    }
  }, []);

  useEffect(() => {
    if (MetaMaskOnboarding.isMetaMaskInstalled()) {
      if (accounts.length > 0) {
        setButtonText(accounts[0]);
        setDisabled(true);
        onboarding?.current?.stopOnboarding();
      } else {
        setButtonText(TEXT_CONNECT);
        setDisabled(false);
      }
    }
  }, [accounts]);

  const onClick = async () => {
    try {
      if (MetaMaskOnboarding.isMetaMaskInstalled()) {
        await window.ethereum
          .request({ method: "eth_requestAccounts" })
          .then((newAccounts) => setAccounts(newAccounts));
      } else {
        onboarding.current.startOnboarding();
      }
    } catch (error) {
      console.error(error);
    }
  };

  return (
    <button
      className='btn btn--small disabled:opacity-50 disabled:cursor-not-allowed truncate w-44'
      disabled={isDisabled}
      onClick={onClick}>
      {buttonText}
    </button>
  );
}
