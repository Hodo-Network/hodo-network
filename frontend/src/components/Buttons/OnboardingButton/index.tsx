import MetaMaskOnboarding from '@metamask/onboarding';
import { useEffect, useRef, useState } from 'react';
// import detectEthereumProvider from "@metamask/detect-provider";
// import { useWeb3React } from "@web3-react/core";
// import { Web3Provider } from "@ethersproject/providers";
import {
  TEXT_CONNECT,
  TEXT_CONNECTED,
  TEXT_INSTALL_METAMASK,
} from '../../../constants/text';
import { PureOnboardingButton } from './OnboardingButton';

const OnboardingButton = () => {
  const [label, setLabel] = useState(TEXT_INSTALL_METAMASK);
  const [disabled, setDisabled] = useState(false);
  const [accounts, setAccounts] = useState([]);
  const onboarding = useRef<MetaMaskOnboarding>();
  // const { library, connector, chainId } = useWeb3React<Web3Provider>();
  // const [provider, setProvider] = useState<any>();

  useEffect(() => {
    if (!onboarding.current) {
      onboarding.current = new MetaMaskOnboarding();
    }
  }, []);

  useEffect(() => {
    if (MetaMaskOnboarding.isMetaMaskInstalled()) {
      if (accounts.length > 0) {
        setLabel(TEXT_CONNECTED);
        setDisabled(true);
        onboarding.current?.stopOnboarding();
      } else {
        setLabel(TEXT_CONNECT);
        setDisabled(false);
      }
    }
  }, [accounts]);

  const handleNewAccounts = (newAccounts: any) => {
    setAccounts(newAccounts);
  };

  const onClick = async () => {
    // const updateProvider = async () => {
    //   const result = await detectEthereumProvider();
    //   setProvider(result);
    // };
    // updateProvider();

    try {
      if (MetaMaskOnboarding.isMetaMaskInstalled()) {
        // @ts-ignore
        await ethereum
          .request({ method: 'eth_requestAccounts' })
          .then(handleNewAccounts);
        // @ts-ignore
        ethereum.on('accountsChanged', handleNewAccounts);
        return () => {
          // @ts-ignore
          ethereum?.off('accountsChanged', handleNewAccounts);
        };
      } else {
        // @ts-ignore
        onboarding.current.startOnboarding();
      }
    } catch (error) {
      console.error(error);
    }
  };

  return (
    <PureOnboardingButton
      label={label}
      disabled={disabled}
      onConnect={onClick}
    />
  );
};

export default OnboardingButton;
