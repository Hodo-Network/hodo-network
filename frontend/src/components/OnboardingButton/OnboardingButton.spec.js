import ReactDOM from "react-dom";
import { cleanup } from "@testing-library/react";
import renderer from "react-test-renderer";
import OnboardingButton from "./index";

afterEach(cleanup);

test("renders without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(<OnboardingButton />, div);
});

test("matches snapshot", () => {
  const tree = renderer.create(<OnboardingButton />).toJSON();
  expect(tree).toMatchSnapshot();
});
