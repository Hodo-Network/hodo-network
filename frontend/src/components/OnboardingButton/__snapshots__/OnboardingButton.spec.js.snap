// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<button
  className="btn disabled:opacity-50 disabled:cursor-not-allowed"
  disabled={false}
  onClick={[Function]}
>
  Install MetaMask
</button>
`;
